### PR TITLE
Remove needless functions from an io.ascii test helper file #15782 issue sloved

### LINKED ISSUE #15782: Remove needless functions from an `io.ascii` test helper file
### What is the problem this feature will solve?

A helper file in `io.ascii` tests contains three functions that serve little purpose: https://github.com/astropy/astropy/blob/fefe2261fcf09751d052d80bd8143e72f18f6aea/astropy/io/ascii/tests/common.py#L27-L37

### Describe the desired outcome

The aforementioned functions should be removed and any tests still using them should be updated accordingly. Pull requests targeting one or more functions are welcome.

Potential contributors can have a look at #16644 that removed the helper functions from one file and use that as an example.

### Additional context

`assert_equal()` and `assert_true()` should be inlined, but `assert_almost_equal(a, b)` should be replaced with [`np.testing.assert_allclose(a, b)`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) instead of `assert np.allclose(a, b)` because the dedicated testing function produces a better error message if the check fails.

### Performed removals

- [ ] `assert_equal()`
- [ ] `assert_almost_equal()`
- [ ] `assert_true()`

### PR DESCRIPTION
**Pull request by Rohit-Pujari**
assert_equal()
 assert_almost_equal()
 assert_true()

### Description

Fixes #15782

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! However, failures are real. Please do try run the test suite locally and see.

https://docs.astropy.org/en/latest/development/testguide.html

https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit

**Comment by Rohit-Pujari** [comment]
image
unable to install repo dependencies

**Comment by Rohit-Pujari** [comment]
i have replaced the functions as  mentioned but unable to run the tests

**Comment by Rohit-Pujari** [comment]
image
unable to update/upgrade astropy to the required version , there is a issue with the asdf-astropy,
i have tried to install the latest astropy version but yet the error persist

**Comment by pllim** [comment]
Two things:

1. You need MSVC to compile C-extensions if you are going to build astropy from source. MSVC installation is separate from `pip`. You need to follow your Windows side instructions. https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B
2. You need to fetch tags from upstream to get `setuptools_scm` to render the version properly. `git fetch upstream --tags`

**Comment by dhomeier** [comment]
Thank you for working on this. 2 general comments:

1. `assert` is a statement, not a function, and thus should be generally written without parentheses (a bit surprised Ruff or black don't catch this).
2. The failed imports show that there are still a number of uses of the removed functions left. You should check for any remaining occurrences e.g. with
`git grep assert_.*equal astropy/io/ascii`

**Comment by dhomeier** [comment]
Your last commit brings in a number of changes inside your `.vs` subdirectory; this does not belong to the repo and you should neither add nor remove anything inside. @pllim is this a standard path for visual studio – should perhaps be added to `.gitignore`?

I suggest you `git reset` to your first commit and continue from there, so none of those files will show up.

**Comment by Rohit-Pujari** [comment]
@dhomeier i have double checked those functions uses and there are none remaining, the numpy.testing.assert_**  is similar to those but i have not touched those

**Comment by dhomeier** [comment]
There are at least 4 imports left, see https://github.com/astropy/astropy/actions/runs/7321280634/job/19941477016#step:10:301

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_cds_header_from_readme.py:47]:
```suggestion
        assert table["loggf"].description == (
            "log10 of the gf value - logarithm base "
            "oscillator strength"
        )
```
This one is a bit tricky, as the string itself needs to be held together by the parentheses and `black` kept from complaining, but this should work.
Again, `assert` itself should never use the `(...)` as you could see here it just adds potential for confusion with constructs that do need them. In particular they should be removed from all the one-line statements!

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_cds_header_from_readme.py:236]:
```suggestion
    assert (
        r.header.cols[10].description
        == "DAOSPEC quality parameter Q (large values are bad)"
    )
```
Here it was probably just `black`'s annoying habit of inserting a trailing comma everywhere that created a spurious tuple here, but otherwise this should work.

**Review by dhomeier — CHANGES_REQUESTED** [review]
Thanks; could you try to at least install the `pre-commit` setup locally as documented in the second link Pey Loan posted above? I _think_ this should function without needing a working compiler installation, and you can then check those warnings and diagnostics like https://results.pre-commit.ci/run/github/2081289/1703509121.Z6WkqI4KQVm9d2te2s3qhg before committing.

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (Rohit-Pujari) [reply astropy/io/ascii/tests/test_cds_header_from_readme.py]: image
it gives error

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (Rohit-Pujari) [reply astropy/io/ascii/tests/test_cds_header_from_readme.py:236]: i have changed these lines should i commit

**Comment by dhomeier** [comment]
Someone familiar with Windows configuration will need to look into your `pre-commit` setup problems.
In the meantime you might try to at least install `black` and `ruff` (through conda or pip) to run the checks manually as described in https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions , i.e. either with `tox -e codestyle`, or if that does not work, `black astropy/io/ascii` and `ruff --fix --show-fixes astropy/io/ascii`.

**Comment by Rohit-Pujari** [comment]
image
image
image

**Comment by dhomeier** [comment]
The tox errors look the same as for pre-commit, or perhaps tox depends on pre-commit working.
But with the fixes from Ruff and black in, should be safe to commit and push the next attempt.

**Review by pllim — CHANGES_REQUESTED** [review]
The introduction of irrelevant IDE files is undesirable.

**Comment by pllim** [comment]
If Windows is giving you too much trouble, please consider using a Linux distro via WSL2.

**Comment by dhomeier** [comment]
> The introduction of irrelevant IDE files is undesirable.

I have already requested clearing those from the branch. This will need squashing anyway – do you know if that will take care of removing those from the history for good?

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15787#issuecomment-1876113916

Indeed squashing would clear them out but they have to disappear from the final diff first for that to work. Unless you remember to use the Squash and Merge button, it is safer to squash it locally and force push so we can have a clean PR history before merge.

If the files accidentally get in `main`, it is going to be a pain because I have to break rules and squash the main branch.

**Comment by dhomeier** [comment]
Yes, agreed this should be clean in the local branch first;  that's why I had also suggested as a possible solution to do a (soft) reset to 4c62f32 or be9bf92 and then redo the commit without the junk files.

**Pull request closed without merging**
